### PR TITLE
Verbosity level and other related changes

### PR DESCRIPTION
**Pull request by HeadLessHUN**
#1949

### CODE REVIEW

File: qCC/ccCommandLineCommands.cpp [hunk 1]
```diff
@@ -7620,3 +7513,28 @@ bool CommandDebugCmdLine::process(ccCommandLineInterface& cmd)
...
+	}
+
+	bool ok = false;
+	int verbosityLevel = cmd.arguments().takeFirst().toInt(&ok);
+	if (!ok) {
+		return cmd.error(QObject::tr("Invalid verbosity level %s").arg(verbosityLevel));
+	}
+	else
+	{
+		cmd.print(QObject::tr("Set verbosity level to %i").arg(verbosityLevel));
```
Comment (dgirardeau) [inline qCC/ccCommandLineCommands.cpp:7535]:
```suggestion
		cmd.print(QObject::tr("Set verbosity level to %1").arg(verbosityLevel));
```

File: qCC/ccCommandLineCommands.cpp [hunk 2]
```diff
@@ -1600,11 +1591,11 @@ bool CommandSubsample::process(ccCommandLineInterface& cmd)
 		for (CLCloudDesc& desc : cmd.clouds())
 		{
 			//calculate octree before subsampling, it is passed to subsampling, so it won't be recalculated there.
-			CCCoreLib::DgmOctree* octree = desc.pc->computeOctree(nullptr, false).data();
+			CCCoreLib::DgmOctree* octree = desc.pc->getOctree().data();
+			if (!octree)
+				octree = desc.pc->computeOctree(nullptr, false).data();
 			if (!octree)
-			{
```
Comment (dgirardeau) [inline qCC/ccCommandLineCommands.cpp:1605]: to be restored

File: qCC/ccCommandLineCommands.cpp [hunk 3]
```diff
@@ -7620,3 +7513,28 @@ bool CommandDebugCmdLine::process(ccCommandLineInterface& cmd)
...
+
+CommandSetVerbosity::CommandSetVerbosity()
+	: ccCommandLineInterface::Command(QObject::tr("Set Verbosity"), COMMAND_VERBOSITY)
+{}
+
+bool CommandSetVerbosity::process(ccCommandLineInterface& cmd)
+{
+	if (cmd.arguments().empty())
+	{
+		return cmd.error(QObject::tr("Missing parameter: verbosity level after: %s").arg(COMMAND_VERBOSITY));
```
Comment (dgirardeau) [inline qCC/ccCommandLineCommands.cpp:7525]:
```suggestion
		return cmd.error(QObject::tr("Missing parameter: verbosity level after: %1").arg(COMMAND_VERBOSITY));
```

File: qCC/ccCommandLineCommands.cpp [hunk 4]
```diff
@@ -7620,3 +7513,28 @@ bool CommandDebugCmdLine::process(ccCommandLineInterface& cmd)
...
+	}
+
+	bool ok = false;
+	int verbosityLevel = cmd.arguments().takeFirst().toInt(&ok);
+	if (!ok) {
```
Comment (dgirardeau) [inline qCC/ccCommandLineCommands.cpp:7530]:
```suggestion
	if (!ok)
	{
```

File: qCC/ccCommandLineCommands.cpp [hunk 5]
```diff
@@ -7620,3 +7513,28 @@ bool CommandDebugCmdLine::process(ccCommandLineInterface& cmd)
...
+		return cmd.error(QObject::tr("Invalid verbosity level %s").arg(verbosityLevel));
```
Comment (dgirardeau) [inline qCC/ccCommandLineCommands.cpp:7531]:
```suggestion
		return cmd.error(QObject::tr("Invalid verbosity level %1").arg(verbosityLevel));
```

File: qCC/main.cpp [hunk 6]
```diff
@@ -116,6 +116,23 @@ int main(int argc, char **argv)
             commandLine = false;
             lastArgumentIndex += 2;
         }
+		if (argumentsLocal8Bit[lastArgumentIndex].toUpper() == "-VERBOSITY")
```
Comment (dgirardeau) [inline qCC/main.cpp:119]: I guess you should check that 'lastArgumentIndex < argumentsLocal8Bit.size()' first?

File: qCC/main.cpp [hunk 7]
```diff
@@ -116,6 +116,23 @@ int main(int argc, char **argv)
...
+		{
+			//remove verified local option
+			argumentsLocal8Bit.removeAt(lastArgumentIndex);
+			bool ok = false;
+			int verbosityLevel = argumentsLocal8Bit[lastArgumentIndex + 1].toInt(&ok);
+			if (!ok)
+			{
+				ccLog::Error(QObject::tr("Invalid verbosity level %s").arg(verbosityLevel));
```
Comment (dgirardeau) [inline qCC/main.cpp:127]:
```suggestion
				ccLog::Error(QObject::tr("Invalid verbosity level %1").arg(verbosityLevel));
```

**Review by dgirardeau — CHANGES_REQUESTED** [review]

**Comment by HeadLessHUN** [comment]
Daniel, i wanted to have higher level for command name print/subprocess time. That is why i used warning, maybe we should/could add a new level instead?:)

**Comment by dgirardeau** [comment]
Yes I realized that. I think we need to add more levels indeed.

Warning and Errors can't be 'ignored' (at least not errors). And they have to be reserved for real warnings and errors.

I think we will have to work a little bit more on that, and sadly we might not be able to add that to the release...

**Comment by HeadLessHUN** [comment]
I was thinking about flipping debug and print as well, and allow debug in release mode as well, just make the standard as default.

Maybe we could add a higher lvl print. and that should be enough?

**Comment by dgirardeau** [comment]
Yes, Debug could become the 'verbose' mode, disabled by default (unless you are in Debug mode).

What was your intent by putting the timing logs as 'warning'? Do you think they are more important than standard messages? Or on the contrary part of the verbose/debug logs?

**Comment by HeadLessHUN** [comment]
well i wanted to have a start/end of a command so that should be higher level i think, At least for me these are more important, then messages inside commands.

**Comment by dgirardeau** [comment]
Ok, so something like that?
Debug = 0 (=Verbose)
Print = 1
Important/High = 2
Warning = 3
Error = 4

**Comment by HeadLessHUN** [comment]
for me that looks fine :) I don't really know if we want to mess with gui mode logs or not... I usually never look the console log in gui mode. So i wouldn't bother. But maybe if we are implementing this, it worth a select menu inside the preferences menu.

**Comment by dgirardeau** [comment]
Yes, for now it's ok to only give this option to the command line users (as in the GUI mode, you can simply hide the console).

I'll try something quick so as to ship that with the release.

**Comment by dgirardeau** [comment]
Ok, here's the new version:
image

I've tested it quickly. You can try it on your side. And we should be good to add it to the release!

File: qCC/ccCommandLineParser.h [hunk 8]
```diff
@@ -55,7 +55,9 @@ class ccCommandLineParser : public ccCommandLineInterface
 	const QStringList& arguments() const override { return m_arguments; }
 	bool registerCommand(Command::Shared command) override;
 	QDialog* widgetParent() override { return m_parentWidget; }
+	void printVerbpse(const QString& message) const override;
```
Comment (HeadLessHUN) [inline qCC/ccCommandLineParser.h:58]: typoo here too

File: libs/CCPluginAPI/include/ccCommandLineInterface.h [hunk 9]
```diff
@@ -295,7 +295,9 @@ class CCPLUGIN_LIB_API ccCommandLineInterface
 public: //logging
 
 	//logging
+	virtual void printVerbpse(const QString& message) const = 0;
```
Comment (HeadLessHUN) [inline libs/CCPluginAPI/include/ccCommandLineInterface.h:298]: typoo
printVerbose

File: qCC/ccCommandLineParser.h (hunk 8 above)
Comment (dgirardeau) [reply qCC/ccCommandLineParser.h:58]: Wow, I was tired 😫

**Comment by HeadLessHUN** [comment]
currently i have build error, but i'm sure it just my environment went wrong somewhere. but i was not able to test these changes yet.

**Comment by dgirardeau** [comment]
Ah ok, keep me updated!

**Comment by HeadLessHUN** [comment]
> Ah ok, keep me updated!

well something is really f*cked up :) my env however is clean and reset to the verbosity_level branch, but they contain changes from the master branch which you pushed yesterday. So there are missing files. bcadd49365e51e75836529fee64c5253e3e6932c

**Comment by dgirardeau** [comment]
Ok I need to merge some stuff in the PythonPlugin first.

**Comment by ryukisai** [comment]
Thanks for this guys! I had not realized it was already implemented.

**Pull request merged**